juju/juju
# fix: process peer integrations last

The linked bug describes issues when removing applications with both normal integrations and peer integrations.

Hooks for some integrations rely on the application data populated by peers. When peer integrations are departed/broken first, these hooks fail, blocking tear-down.

While we generally make no guarantees to charm authors about ordering, it is reasonable to process existing integrations peer-last to prevent this situation. 

Hooks for creation and joining are unaffected.

## QA steps

- Bootstrap and add a model.
- Make sure logging config includes `unit=DEBUG`.
- `juju deploy postgresql -n 2`.
- `juju deploy postgresql pgsink`.
- `juju relate postgresql:replication-offer pgsink`.
- Wait for quiescence.
- `Juju remove-appllication postgresql`, and wait for it to be gone.
- `juju debug-log -m work --replay|grep -E 'postgresql-0.*via hook'|less`
- You will see that the peer relations are departed after the replication relation.
```
unit-postgresql-0: 14:31:26 INFO juju.worker.uniter.operation ran "replication-offer-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-0: 14:31:31 INFO juju.worker.uniter.operation ran "pgdata-storage-detaching" hook (via hook dispatching script: dispatch)
unit-postgresql-0: 14:31:33 INFO juju.worker.uniter.operation ran "replication-offer-relation-broken" hook (via hook dispatching script: dispatch)
unit-postgresql-0: 14:31:34 INFO juju.worker.uniter.operation ran "database-peers-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-0: 14:31:36 INFO juju.worker.uniter.operation ran "restart-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-0: 14:31:37 INFO juju.worker.uniter.operation ran "upgrade-relation-departed" hook (via hook dispatching script: dispatch)
```

## Documentation changes

None.

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/1998282

**Jira card:** [JUJU-6705](https://warthogs.atlassian.net/browse/JUJU-6705)
